Firmware/Marlin_main.cpp

```diff
@@ -10263,6 +10264,15 @@ static void temp_compensation_start() {
 	custom_message_state = 0;
 }
 
+static void temp_compensation_stop()
+{
+    //unretract (after PINDA preheat retraction)
+    if (((int)degHotend(active_extruder) > extrude_min_temp) && eeprom_read_byte((unsigned char *)EEPROM_TEMP_CAL_ACTIVE) && calibration_status_pinda() && (target_temperature_bed >= 50)) {
+        current_position[E_AXIS] += default_retraction;
```

Comment:
There is still another potential race here if the extruder wasn't hot enough during temp_compensation_start() then the retraction is never done. Also similarly, the conditions done here do not match the conditions of the retract anyway.

I would change temp_compensation_start() to return a boolean indicating if the retract is done, then just unretract instead of doing all this checking.